packages/flutter/lib/src/foundation/memory_allocations.dart

```diff
@@ -0,0 +1,278 @@
+  Map<Object, Map<String, dynamic>> toDart();
+}
+
+/// A listener of [ObjectEvent].
+typedef ObjectEventListener = void Function(ObjectEvent);
+
+/// An event that describes creation of an object.
+class ObjectCreated extends ObjectEvent {
+  /// Creates an instance of [ObjectCreated].
+  ObjectCreated({
+    required this.library,
+    required this.className,
+    required super.object,
+  });
+
+  /// Name of the instrumented library.
+  final String library;
+
+  /// Name of the instrumented class.
+  final String className;
+
+  @override
+  Map<Object, Map<String, dynamic>> toDart() =>
+    <Object, Map<String, dynamic>>{object: <String, dynamic>{
+      FieldNames.labraryName: library,
+      FieldNames.className: className,
+      FieldNames.eventType: 'created',
+    }};
+}
+
+/// An event that describes disposal of an object.
+class ObjectDisposed extends ObjectEvent {
+  /// Creates an instance of [ObjectDisposed].
+  ObjectDisposed({
+    required super.object,
+  });
+
+  @override
+  Map<Object, Map<String, dynamic>> toDart() =>
+    <Object, Map<String, dynamic>>{object: <String, dynamic>{
+      FieldNames.eventType: 'disposed',
+    }};
+}
+
+/// An interface for listening to object lyfecycle events.
+///
+/// [MemoryAllocations] already listens to creation and disposal events
+/// for disposable objects in Flutter Framework.
+/// You can dispatch events for other objects by invoking
+/// [MemoryAllocations.dispatchObjectEvent].
+///
+/// The class is optimized for massive event flow and small number of
+/// added or removed listeners.
+class MemoryAllocations {
+  MemoryAllocations._();
+
+  /// The shared instance of [MemoryAllocations].
+  static final MemoryAllocations instance = MemoryAllocations._();
+
+  /// List of listeners.
+  ///
+  /// The elements are nullable, because the listeners should be removable
+  /// while iterating through the list.
+  List<ObjectEventListener?>? _listeners;
+
+  /// Register a listener that is called every time an object event is
+  /// dispatched.
+  ///
+  /// Listeners can be removed with [removeListener].
+  void addListener(ObjectEventListener listener){
+    if (_listeners == null) {
+      _listeners = <ObjectEventListener?>[];
+      _subscribeToSdkObjects();
+    }
+    _listeners!.add(listener);
+  }
+
+  /// Number of active notification loops.
+  ///
+  /// When equal to zero, we can delete listeners from the list,
+  /// otherwize should null them.
+  int _activeDispatchLoops = 0;
+
+  /// If true, listeners were nulled by [removeListener].
+  bool _listenersContainNulls = false;
+
+  /// Stop calling the given listener every time an object event is
+  /// dispatched.
+  ///
+  /// Listeners can be added with [addListener].
+  void removeListener(ObjectEventListener listener){
+    final List<ObjectEventListener?>? listeners = _listeners;
+    if (listeners == null) {
+      return;
+    }
+
+    if (_activeDispatchLoops > 0) {
```

Comment:
For my own understanding, why is the removal/notification logic so different from ChangeNotifier? I wish we could just reuse that logic instead of reinventing everything.

Reply:
ChangeNotifier is different, because it is optimized for adding/removing listeners, I think.